xchange-bitstamp/src/main/java/info/bitrich/xchangestream/bitstamp/v2/BitstampStreamingMarketDataService.java

```diff
@@ -0,0 +1,77 @@
+package info.bitrich.xchangestream.bitstamp.v2;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import info.bitrich.xchangestream.bitstamp.dto.BitstampOrderBook;
+import info.bitrich.xchangestream.bitstamp.dto.BitstampWebSocketTransaction;
+import info.bitrich.xchangestream.core.StreamingMarketDataService;
+import info.bitrich.xchangestream.service.netty.StreamingObjectMapperHelper;
+import io.reactivex.Observable;
+import org.knowm.xchange.bitstamp.BitstampAdapters;
+import org.knowm.xchange.currency.CurrencyPair;
+import org.knowm.xchange.dto.marketdata.OrderBook;
+import org.knowm.xchange.dto.marketdata.Ticker;
+import org.knowm.xchange.dto.marketdata.Trade;
+import org.knowm.xchange.exceptions.NotAvailableFromExchangeException;
+
+import java.util.Date;
+
+/**
+ * Bitstamp WebSocket V2 Streaming Market Data Service implementation
+ * Created by Pavel Chertalev on 15.03.2018.
+ */
+public class BitstampStreamingMarketDataService implements StreamingMarketDataService {
+    private final BitstampStreamingService service;
+
+    public BitstampStreamingMarketDataService(BitstampStreamingService service) {
+        this.service = service;
+    }
+
+    public Observable<OrderBook> getDifferentialOrderBook(CurrencyPair currencyPair, Object... args) {
+        return getOrderBook("diff_order_book", currencyPair, args);
+    }
+
+    @Override
+    public Observable<OrderBook> getOrderBook(CurrencyPair currencyPair, Object... args) {
+        return getOrderBook("order_book", currencyPair, args);
+    }
+
+    private Observable<OrderBook> getOrderBook(String channelPrefix, CurrencyPair currencyPair, Object... args) {
+        String channelName = channelPrefix + getChannelPostfix(currencyPair);
+
+        return service.subscribeChannel(channelName, "data")
+                .map(s -> {
+                    ObjectMapper mapper = StreamingObjectMapperHelper.getObjectMapper();
+                    BitstampOrderBook orderBook = mapper.treeToValue(s.get("data"), BitstampOrderBook.class);
+                    org.knowm.xchange.bitstamp.dto.marketdata.BitstampOrderBook bitstampOrderBook =
+                            new org.knowm.xchange.bitstamp.dto.marketdata.BitstampOrderBook(
+                                    orderBook.getTimestamp(),
+                                    orderBook.getBids(),
+                                    orderBook.getAsks());
+                    return BitstampAdapters.adaptOrderBook(bitstampOrderBook, currencyPair);
+                });
+    }
+
+    @Override
+    public Observable<Ticker> getTicker(CurrencyPair currencyPair, Object... args) {
+        // BitStamp has no live ticker, only trades.
```

Comment:
Easily implemented by mapping the trade stream, though?

One for another day?

Reply:
Bitstamp does not have ticker stream channel. 
XChange ticker contains the following data:

```
public final class Ticker implements Serializable {
    private final CurrencyPair currencyPair;
    private final BigDecimal open;
    private final BigDecimal last;
    private final BigDecimal bid;
    private final BigDecimal ask;
    private final BigDecimal high;
    private final BigDecimal low;
    private final BigDecimal vwap;
    private final BigDecimal volume;
    private final BigDecimal quoteVolume;
    private final Date timestamp;
    private final BigDecimal bidSize;
    private final BigDecimal askSize;
    ...
}
```
Is it possible to calculate accurate ticker values using trades stream? Anyway old V1 version does not implement one as well. Looks like Bitstamp wants rest call for ticker information :)

Reply:
You don't have to populate all the values. Lots of exchanges don't. There's a `Builder` that permits missing values.

Bid, last and ask are the only ones I'd expect in all cases. You can merge the order book and trade streams to produce those (with a few `AtomicReference`s).

Don't worry about it unless you feel like it though; this can easily be added later.

Reply:
ok